add description of acrablock container, encryption/decryption processes (#41)
cossacklabs/product-docs: acra/acrablock/_index.md

---
title: AcraBlock
bookCollapseSection: true
---

# AcraBlock

## Container structure

> AcraBlock is a container format. To generate AcraBlocks Acra-Server uses symmetric keys generated for every ClientID/ZoneID.

Reviewer: ```suggestion
AcraBlock is a container format. Acra uses an envelope encryption strategy: plaintext data is encrypted using data encryption key (DEK), and then DEK is encrypted using key encryption key (KEK). AcraBlock uses AES-GCM for both encryption procedures. 

AcraBlock supports key rotation: it's possible to rotate only key encryption key without re-encrypting the data (known as "key rotation without data re-encryption"), or re-encrypt data and rotate both keys ("key rotation with data re-encryption").

To generate AcraBlocks Acra-Server uses symmetric keys generated for every ClientID/ZoneID.
```

* `Key_Encryption_Key_ID[2]` - this field store KEK's identifier. Main goal of this ID is to simplify search of KEK in a keychain related to specific ClientID/ZoneID. Now Acra-Server generates ID by hashing with `SHA256(KEK || ClientID/ZoneID)` and uses first 2 bytes as ID.
* `Data_Encryption_Backend_Identifier[1]` - identifier of encryption backend that used to encrypt plaintext. For now AcraBlock supports only one backend that uses Themis Secure Cell with Seal mode with identifier `0`. Same as for DEK encryption and may be changed too.
* `Data_Encryption_Key_Length[2]` - 2 bytes of DEK's length in Little-Endian byte order. Key size should be synchronized with data encryption backend.
* `Encrypted_Data_Encryption_Key[*]` - encrypted DEK of size `Data_Encryption_Key_Length`, used to encrypt data and data encryption backend. For each new AcraBlock Acra-Server generates new random DEK and encrypts with DEK encryption backend and Context (ClientID/ZoneID).
* `Encrypted_Data[*]` encrypted data with data encryption backend and random DEK.

## Generating

> To generate AcraBlock in transparent mode Acra-Server do next steps:

Reviewer: ```suggestion
To generate AcraBlock in transparent mode Acra-Server does the next steps:
```

> * Generates new random DEK for default data encryption backend. New it is 32 bytes.

Reviewer: ```suggestion
* Generates new random DEK for default data encryption backend. DEK length is 32 bytes.
```

> * Encrypts plaintext by this key and Context of specified ClientID/ZoneID using default backend for data encryption.

Reviewer: ```suggestion
* Encrypts plaintext by DEK and Context of specified ClientID/ZoneID using default backend for data encryption.
```

> * Erases/fills with zeros memory area with the plaintext

Reviewer: ```suggestion
* Securely cleans up memory of plaintext data (erases/fills with zeros).
```

* Encrypts DEK by KEK from KeyStore for specified ClientID/ZoneID and use it as Context

> * Erases/fills with zeros memory area with the DEK

Reviewer: ```suggestion
* Securely cleans up memory of DEK (erases/fills with zeros).
```

> * Pack together Begin_Tag, length of all other parts, DEK encryption backend identifier, KEK identifier, DEK encryption backend identifier, encrypted DEK and ciphertext

Reviewer: ```suggestion
* Forms a container: packs together Begin_Tag, length of all other parts, DEK encryption backend identifier, KEK identifier, DEK encryption backend identifier, encrypted DEK and ciphertext.
```
